### PR TITLE
Add ErrorHandler subscriber

### LINKED ISSUE #66: errorResposnes from operation is never used
Hi,

In the Operation descriptor, there is a "errorResponses" hash that is supposed to be used. While it's read by the operation, it seems to be never used elsewhere in the code. As a consequence, user defined exception are never triggered.

Did I miss something or is the doc outdated? :)

Thanks!

### PR DESCRIPTION
**Pull request by bakura10**
Hi,

This PR fixes #66.

It adds an optional subscriber that reads the errorResponses array and throw exception.

A few things:

* You must register the listener in the client
* Your exceptions must extend CommandException, because any other exception is wrapped automatically around a CommandException (which is still hell annoying).

### CODE REVIEW

File: src/Subscriber/HandleErrorResponse.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
+<?php
+/*
```
Comment (GrahamCampbell) [inline src/Subscriber/HandleErrorResponse.php]: please remove this

File: tests/Asset/Exception/CustomCommandException.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<?php
+/*
```
Comment (GrahamCampbell) [inline tests/Asset/Exception/CustomCommandException.php]: please remove this

File: tests/Asset/Exception/OtherCustomCommandException.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
+<?php
+/*
```
Comment (GrahamCampbell) [inline tests/Asset/Exception/OtherCustomCommandException.php]: please remove this

File: tests/Asset/Exception/OtherCustomCommandException.php [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * and is licensed under the MIT license.
+ */
+
+namespace GuzzleHttp\Tests\Command\Guzzle\Asset\Exception;
+
+use GuzzleHttp\Command\Exception\CommandException;
+
+class OtherCustomCommandException extends CommandException
+{
+}
```
Comment (GrahamCampbell) [inline tests/Asset/Exception/OtherCustomCommandException.php]: all files must end in a single new line

**Comment by bakura10** [comment]
Done.

File: src/Subscriber/HandleErrorResponse.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
+<?php
+namespace GuzzleHttp\Command\Guzzle\Subscriber;
+
+use GuzzleHttp\Command\Event\ProcessEvent;
+use GuzzleHttp\Event\SubscriberInterface;
+
+/**
+ * Subscriber that can be used when filling the "errorResponses" property for commands. In order for the
```
Comment (GrahamCampbell) [inline src/Subscriber/HandleErrorResponse.php]: This is not the correct description format.

Comment (GrahamCampbell) [inline src/Subscriber/HandleErrorResponse.php]: You want it like this:

```php
/**
 * Subscriber that can be used when filling the "errorResponses" property for commands.
 *
 * In order for the...
```

**Comment by bakura10** [comment]
This approach is actually less flexible than what we had in Guzzle 3. Previously we had a "fromCommand" static method that allowed to create an exception based on the response. I'm encountering the need to do that in my Stripe client.

I'll try to see how I can change that.

**Comment by bakura10** [comment]
Maybe I could add a "CommandExceptionFactory" abstract class with a simple "factory" static method that would accept a CommandTransaction and return an exception. What do you think?

File: src/Subscriber/HandleErrorResponse.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
+<?php
+namespace GuzzleHttp\Command\Guzzle\Subscriber;
+
+use GuzzleHttp\Command\Event\ProcessEvent;
+use GuzzleHttp\Event\SubscriberInterface;
+
+/**
+ * Subscriber that reads the "errorResponses" from commands, and trigger appropriate exceptions
```
Comment (GrahamCampbell) [inline src/Subscriber/HandleErrorResponse.php]: You need a fullstop at the end of this line.

File: src/Subscriber/HandleErrorResponse.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * In order for the exception to be properly triggered, all your exceptions must be instance
+ * of "GuzzleHttp\Command\Exception\CommandException". If that's not the case, your exceptions will be wrapped
+ * around a CommandException
```
Comment (GrahamCampbell) [inline src/Subscriber/HandleErrorResponse.php]: You need a fullstop at the end of this line.

**Comment by mtdowling** [comment]
This is cool, and seems simple enough. I'll merge this if you think it's ready.

> Maybe I could add a "CommandExceptionFactory" abstract class with a simple "factory" static method that would accept a CommandTransaction and return an exception. What do you think?

I'm not a fan of this idea. The point of the event system was to support doing anything from retrying a command to throwing a custom exception (as you've demonstrated with this PR, it's possible). The event system is how a library author should implement throwing custom exceptions.

**Comment by bakura10** [comment]
Hi,

The use case I have is that, for instance, a 400 error have an additional error code that should trigger a precise exception. We have two options:

1. We consider that this is too specific enough, and encourage to use a custom listener.
2. We reintroduce back a kind of factory.

If you think 1 is better, then you can merge :).

**Pull request merged**